fundamental-concepts/data-structures-stdlib/7_lists.md

```diff
@@ -0,0 +1,112 @@
+
+<pre class=file data-target="clipboard">
+data Person = Person {
+  name: Text,
+  addresses: [Address],
+  age: Int
+}
+</pre>
+
+We need to adapt the `testPerson` function as well:
+
+<pre class=file data-target="clipboard">
+testPerson = Person {name = "Alice", addresses = [Address {street = "RabbitStreet", city = "QueenOfHearts", country = "Wonderland"}], age = 7}
+</pre>
+
+The `DA.List` module and the `DA.Prelude` (which is always imported) contains many important
+algorithms to work with lists. Here are just a few of the most common and important:
+
+- (++): [a] -> [a] -> [a]
+  Concatenate two lists.
+- (::): a -> [a] -> [a]
+  Add an element to the beginning of a list.
+- sort: Ord a => [a] -> [a]
+  Stable sorting of lists with elements that can be compared.
+  `sort [3, 2, 1] -- [1, 2, 3]`
+- dedup: Ord a => [a] -> [a]
+  Removes duplicates from a list.
+  `dedup [3, 3, 2, 1]` -- [3, 2, 1]`
+- takeWhile: (a -> Bool) -> [a] -> [a]
+  Take all elements from the beginning of a list as long as the given predicate is satisfied.
+  `takeWhile (> 3) [5, 4, 3, 2, 1] -- [5, 4]`
+- head: [a] -> a
+  Take the first element from a list. Be sure to assert that the list isn't empty, otherwise you'll
+  get a runtime error.
+  `head [1, 2, 3] -- 1`
+- last: [a] -> a
+  Extract the last element from a list. Again, make sure that the list is not empty.
+  `last [1, 2, 3] -- 3`
+- delete: a -> [a] -> a
+  Delete the first occurence of the given element from a list if it is present.
+  delete 1 [1, 2, 3] -- [2, 3]
+- `(\\): Eq a => [a] -> [a] -> [a]`
+  List difference.
+  `[1, 2, 3, 4, 5, 6] \\ [1, 3, 5] -- [2, 4, 6]`
+- foldl: (b -> a -> b) -> b -> [a] -> b
+  This is a left associative fold over a list.
+  `foldl (\x y -> x + y) 0 [1, 2, 3] -- 6`
+
+Be sure to check the documentation of [DA.List](https://docs.daml.com/daml/stdlib/DA-List.html) and
+[Prelude](https://docs.daml.com/daml/stdlib/Prelude.html) before trying to implement your own list
+algorithms, there's a fair chance it's already there!
+
+Remember that you can also search the [docs](https://docs.daml.com) with queries for types. Say
+you're looking for the `head` function but you can't remember its name, but only its type: `[a] ->a`.
+Searching for `[a] -> a` will indeed give `head` as first result, followed by `last` and other
+functions with the same or similar type.
+
+![search_head](assets/search_head.png)
+
+>> How would you search for the sort function? <<
+() `a -> [a]`
+(*) `[a] -> [a]`
+() `a -> a`
+() `[a] -> a`
+
+If you're coming from an imperative language like Java or C/C++, you might be wondering how to write
+a `for` loop to iterate over a list. There are no loop constructs in DAML! Instead, you can map over
+a list with `map: (a -> b) -> [a] -> [b]` and accumulate local state with  folding functions like
+`foldl: (b -> a -> b) -> b -> [a]-> b`. Take a look at the following `for`-loop in Java:
```

Comment:
```suggestion
`foldl: (b -> a -> b) -> b -> [a]-> b`. As a reminder `foldl f i xs` performs a left fold over the list `xs` using the function `f`, using the starting value `i`. Take a look at the following `for`-loop in Java:
```
I'm still struggling with the notation used in the docs for describing functions, had to go there and read what `foldl` does. Adding this description/reminder might help noobs (feel free to reword it).

Reply:
extended the description in the above list.